src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
```

Comment:
> basically removing a document will clear up all diagnostic associated with that document.

How do you know this?  It's just an interface.  How does this code here know that that's how that interface method will be implemented?

> project one is not called since we never asked project to be analyzed.

How do you know that the analyzer won't *still* have cached data with the project?

--
Let me restate my concern:  your code here probably works.  However, it works *precisely* because you know how this method has been implemented in every case.

That's an *extremely* brittle way of coding this up.  You are depending on all that code working in that manner, and any small change there would break this assumption here and could lead to leaks again.

If you want the analyzer to clear cached data, then you should have an API for that on it that is clearly named and clearly documented to do precisely that.  Then, it's clear for all implementers what they have to do and what granularity they have to work at.

This code should be written as if the PreviewSolutionCrawlerRegistrationService has no idea how the IIncrementalAnalyzers are written.  We should not have disparate components of the system written such that they take a brittle dependency on exactly how things are being implemented.